File: frontend/src/stores/alarm.ts
import { defineStore } from "pinia";
import { ref, computed } from "vue";
import api from "@/api/client";

export interface Alarm {
  id: number;
  device_id: number;
  alarm_type: "DEVICE_DOWN" | "DEVICE_UP" | "INTERFACE_DOWN" | "INTERFACE_UP" | "POLL_TIMEOUT";
  severity: "CRITICAL" | "WARNING" | "INFO";
  message: string;
  status: "OPEN" | "ACKNOWLEDGED" | "RESOLVED";
  created_at: string;
  resolved_at: string | null;
}

export const useAlarmStore = defineStore("alarm", () => {
  const alarms = ref<Alarm[]>([]);
  const polling = ref<ReturnType<typeof setInterval> | null>(null);

  const openAlarms = computed(() => alarms.value.filter((a) => a.status === "OPEN"));
  const openCount = computed(() => openAlarms.value.length);

  async function fetchAlarms(status?: string) {
    const { data } = await api.get("/alarms", { params: status ? { status } : {} });
    alarms.value = data;
  }

  async function acknowledgeAlarm(id: number) {
    const { data } = await api.post(`/alarms/${id}/acknowledge`);
    const idx = alarms.value.findIndex((a) => a.id === id);
    if (idx !== -1) alarms.value[idx] = data;
  }

  async function addAction(alarmId: number, action_type: string, note?: string) {
    await api.post(`/alarms/${alarmId}/actions`, { action_type, note });
  }

  function startPolling(intervalMs = 3000) {
    if (polling.value) return;
    fetchAlarms("OPEN");
    polling.value = setInterval(() => fetchAlarms("OPEN"), intervalMs);
  }

  function stopPolling() {
    if (polling.value) {
      clearInterval(polling.value);
      polling.value = null;
    }
  }

  return {
    alarms, openAlarms, openCount,
    fetchAlarms, acknowledgeAlarm, addAction,
    startPolling, stopPolling,
  };
});

File: backend/app/schemas/alarm.py
from datetime import datetime

from pydantic import BaseModel

from app.models.alarm import AlarmSeverity, AlarmStatus, AlarmType


class AlarmResponse(BaseModel):
    id: int
    device_id: int
    alarm_type: AlarmType
    severity: AlarmSeverity
    message: str
    status: AlarmStatus
    created_at: datetime
    resolved_at: datetime | None
    actions: list["AlarmActionResponse"] = []

    model_config = {"from_attributes": True}


class AlarmActionCreate(BaseModel):
    action_type: str
    note: str | None = None


class AlarmActionResponse(BaseModel):
    id: int
    alarm_id: int
    user_id: int
    action_type: str
    note: str | None
    created_at: datetime

    model_config = {"from_attributes": True}


AlarmResponse.model_rebuild()
